### DISCUSSION TITLE
v0.14.0: How to broadcast to all sockets

### DISCUSSION THREAD

**Discussion by christian3042** [Totodore/socketioxide#351] [Q&A] [answered]
I read the discussion in https://github.com/Totodore/socketioxide/discussions/323 and it was discussed, that broadcast() will emit to the sender too. In the documentation of v0.14.0 it says "Broadcasts to all sockets without any filtering (except the current socket)."
For my purpose I need a broadcast() to send to all sockets, including the sender. In the official socket.io documentation (https://socket.io/docs/v3/broadcasting-events/) there are different kinds of broadcasts:
- To all connected clients
- To all connected clients except the sender

As I'm new to Rust and socket.io I was digging through the examples, and also couldn't find an example for sending to all connected sockets, except from working around with (https://github.com/Totodore/socketioxide/blob/main/examples/private-messaging/src/handlers.rs):

    s.emit("users", [users]).unwrap();
    let res = UserConnectedRes::new(&session, vec![]);
    s.broadcast().emit("user connected", res).unwrap();

Is there a reason to not have a brodcastAll method implemented, or did I miss it? Also the README could have the basic emit operations explained. I could also help with improving the documentation.

**Comment by Totodore** [comment]
You can use the [io.emit](https://docs.rs/socketioxide/latest/socketioxide/struct.SocketIo.html#method.emit) method. This will emit on all sockets in the "/" namespace, you can also add more filter by using operators as usual [(`of`/`within`/`to`,...)](https://docs.rs/socketioxide/latest/socketioxide/operators/index.html).

I agree that it could be specified in the emit section of the documentation. Feel free to contribute if you want:
 https://docs.rs/socketioxide/latest/socketioxide/#emiting-data

**Comment by christian3042** [comment]
In my Axum application I have in the main method the following code to initialize the Socket.io:

    let (websocketLayer, io) = SocketIo::builder().build_layer();
    io.ns("/", on_connect);

my on_connect handler looks as follows:

```
pub fn on_connect(socket: SocketRef) {
    socket.on(
        "event",
        |socket: SocketRef| {
            socket.emit("event-res", "ok").ok();
            socket.broadcast().emit("event-res", "ok").ok();
        },
    );
}
```

When I have two clients connected to my server, the code above leads to the expected behavior. Every client receives a response once, independent of who has sent the event. So in my case, the first emit does not get sent to all the clients connected to the server. Not sure if I understood it correctly, or if I'm doing something wrong.

**Reply by Totodore** [reply, marked as answer]
You can do that or use `io.emit` as I mentioned above.

**Reply by christian3042** [reply]
Thanks for explaining! I didn't know that SocketRef emit is different than the io.emit.
